### ISSUE TITLE
incorrect core counts from 'flux ion-R encode' for heterogeneous cluster

### ISSUE THREAD

**Issue by ryanday36** [flux-framework/flux-sched#912]
In some cases, a cluster may have, for example, login nodes with fewer cores than compute nodes. Consider a cluster 'foo' with a management node 'foo2' with 2 cores, a login node 'foo3' with 1 core, and two compute nodes foo[1,4] with 2 cores that is encoded as:

```
[day36@corona81:flux]$ (flux R encode -r 0 -c 0-1 -H foo2 && flux R encode -r 1 -c 0 -H foo3 && flux R encode -r 2-3 -c 0-1 -H foo[1,4]) | flux R append | flux ion-R encode
{"version": 1, "execution": {"R_lite": [{"rank": "0,2-3", "children": {"core": "0-1"}}, {"rank": "1", "children": {"core": "0"}}], "starttime": 0.0, "expiration": 0.0, "nodelist": ["foo[2-3,1,4]"]}, "scheduling": {"graph": {"directed": true, "nodes": [{"id": "0", "metadata": {"type": "cluster", "basename": "cluster", "name": "cluster0", "id": 0, "uniq_id": 0, "rank": -1, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0"}}}, {"id": "1", "metadata": {"type": "node", "basename": "node", "name": "foo2", "id": 2, "uniq_id": 1, "rank": 0, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo2"}}}, {"id": "2", "metadata": {"type": "core", "basename": "core", "name": "core0", "id": 0, "uniq_id": 2, "rank": 0, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo2/core0"}}}, {"id": "3", "metadata": {"type": "core", "basename": "core", "name": "core1", "id": 1, "uniq_id": 3, "rank": 0, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo2/core1"}}}, {"id": "4", "metadata": {"type": "node", "basename": "node", "name": "foo3", "id": 3, "uniq_id": 4, "rank": 2, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo3"}}}, {"id": "5", "metadata": {"type": "core", "basename": "core", "name": "core0", "id": 0, "uniq_id": 5, "rank": 2, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo3/core0"}}}, {"id": "6", "metadata": {"type": "core", "basename": "core", "name": "core1", "id": 1, "uniq_id": 6, "rank": 2, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo3/core1"}}}, {"id": "7", "metadata": {"type": "node", "basename": "node", "name": "foo1", "id": 1, "uniq_id": 7, "rank": 3, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo1"}}}, {"id": "8", "metadata": {"type": "core", "basename": "core", "name": "core0", "id": 0, "uniq_id": 8, "rank": 3, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo1/core0"}}}, {"id": "9", "metadata": {"type": "core", "basename": "core", "name": "core1", "id": 1, "uniq_id": 9, "rank": 3, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo1/core1"}}}, {"id": "10", "metadata": {"type": "node", "basename": "node", "name": "foo4", "id": 4, "uniq_id": 10, "rank": 1, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo4"}}}, {"id": "11", "metadata": {"type": "core", "basename": "core", "name": "core0", "id": 0, "uniq_id": 11, "rank": 1, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/cluster0/foo4/core0"}}}], "edges": [{"source": "0", "target": "1", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "1", "target": "2", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "1", "target": "3", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "0", "target": "4", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "4", "target": "5", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "4", "target": "6", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "0", "target": "7", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "7", "target": "8", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "7", "target": "9", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "0", "target": "10", "directed": true, "metadata": {"name": {"containment": "contains"}}}, {"source": "10", "target": "11", "directed": true, "metadata": {"name": {"containment": "contains"}}}]}}}
[day36@corona81:flux]$
```

In this encoding, the 'flux R encode' has listed ranks 0,2-3 has having 2 cores and rank 1 as having 1 core, but 'flux ion-R encode' has assigned two cores to flux[1-3] / ranks 0-2 and only one core to flux4 / rank 3.

**Comment by ryanday36** [comment]
I should also note that one practical effect of this is errors / warnings at job allocation. e.g. on corona, we have three login nodes with 24 cores vs. compute nodes with 48 cores. If I list those login nodes before the compute nodes when generating R, I see things like:
```
[day36@corona82:~]$ flux mini alloc -n 1
2022-03-02T17:21:43.267171Z resource.err[0]: verify: rank 0 got hostname 'corona160', expected 'corona157'
[day36@corona160:~]$
```
Listing the login nodes last when generating R, so that they get the correct core counts from `flux ion-R encode` eliminates this error/warning message.

**Comment by grondo** [comment]
Oh, you know there was also an existing issue that may be related: #893. I had thought this was fixed but the issue is still open, so now I'm uncertain...

**Comment by ryanday36** [comment]
So, maybe too many moving parts here. The error from `flux mini alloc` doesn't actually go away unless I also put the login nodes last in the `hosts` section of `bootstrap.toml` (so it matches the order in `R`). There's also some weirdness in the output of `flux resource list` that I think probably deserves its own issue. When I list the three login nodes before the compute nodes in `bootstrap.toml`, the numbering in 'NODELIST' for the 'down' nodes appears to be off by 3, but the 'free' node numbering looks right. When I list the three login nodes last in `bootstrap.toml`, the 'down' node numbering is right, but the 'free' numbering is off by 3.

**Comment by dongahn** [comment]
Could be the same bug as https://github.com/flux-framework/flux-sched/issues/893

**Comment by grondo** [comment]
Yes, the `hostlist` in Rv1 is in strict rank order.

**Comment by grondo** [comment]
@dongahn,  is #893 still open? If it is fixed in 0.21.0, then we should have a fix for @ryanday36 soon. If it is not fixed, then do you have an estimated time to land a fix? Perhaps we should wait on 0.21.1 until #893 is fixed.

**Comment by grondo** [comment]
> hostlist lists hostnames in rank order. Correct @grondo?

Just to clarify, the way to use the `hostlist` in R is to encode it as a `struct hostlist` or `Hostlist` in Python, and then to translate rank to hostname just do `hostlist_nth (hostlist, rank)` or `hostlist[rank]` in python.

**Comment by grondo** [comment]
I could probably take a look at modification of `flux-ion R`, but unfortunately I fear I may not be familiar enough with Fluxion to help with the R reader.

**Comment by dongahn** [comment]
I will take a quick look. Let me see if I can carve out some time tonight. Cannot be difficult (famous last word :-)

**Comment by grondo** [comment]
> Let me see if I can carve out some time tonight. Cannot be difficult (famous last word :-)

Hi @dongahn, it wasn't clear to me if you wanted me to take a stab at fixing this issue, or were you going to work on both this and #893? Happy to help in any way.

**Comment by dongahn** [comment]
OK. I have a patch for this `flux ion-R` issue along with some test cases. I will put together a PR for this shortly. 

```diff
diff --git a/src/python/fluxion/resourcegraph/V1.py b/src/python/fluxion/resourcegraph/V1.py
index ccacd67d..3f0325d5 100644
--- a/src/python/fluxion/resourcegraph/V1.py
+++ b/src/python/fluxion/resourcegraph/V1.py
@@ -141,14 +141,16 @@ class FluxionResourceGraphV1(Graph):
         edg = FluxionResourceRelationshipV1(ppid, vtx.get_id())
         self._add_and_tick_uniq_id(vtx, edg)
 
-    def _encode_rank(self, ppid, rank, children, hList, hIndex):
-        hPath = f"/cluster0/{hList[hIndex]}"
-        iden = self._extract_id_from_hn(hList[hIndex])
+    def _encode_rank(self, ppid, rank, children, hList, rdict):
+        if rdict[rank] >= len(hList):
+            raise Exception(f"nodelist doesn't include node for rank={rank}")
+        hPath = f"/cluster0/{hList[rdict[rank]]}"
+        iden = self._extract_id_from_hn(hList[rdict[rank]])
         vtx = FluxionResourcePoolV1(
             self._uniqId,
             "node",
             "node",
-            hList[hIndex],
+            hList[rdict[rank]],
             iden,
             self._uniqId,
             rank,
@@ -163,14 +165,11 @@ class FluxionResourceGraphV1(Graph):
             for i in IDset(val):
                 self._encode_child(vtx.get_id(), hPath, rank, str(key), i)
 
-    def _encode_rlite(self, ppid, entry, hList, hIndex):
+    def _encode_rlite(self, ppid, entry, hList, rdict):
         for rank in list(IDset(entry["rank"])):
-            hIndex += 1
-            self._encode_rank(ppid, rank, entry["children"], hList, hIndex)
-        return hIndex
+            self._encode_rank(ppid, rank, entry["children"], hList, rdict)
 
     def _encode(self):
-        hIndex = -1
         hList = Hostlist(self._rv1NoSched["execution"]["nodelist"])
         vtx = FluxionResourcePoolV1(
             self._uniqId,
@@ -186,5 +185,13 @@ class FluxionResourceGraphV1(Graph):
             "/cluster0",
         )
         self._add_and_tick_uniq_id(vtx, None)
+        i = 0
+        rdict = {}
         for entry in self._rv1NoSched["execution"]["R_lite"]:
-            hIndex = self._encode_rlite(vtx.get_id(), entry, hList, hIndex)
+            for rank in list(IDset(entry["rank"])):
+                if rank in rdict:
+                    raise Exception(f"R_lite: rank={rank} found again!")
+                rdict[rank] = i
+                i += 1
+        for entry in self._rv1NoSched["execution"]["R_lite"]:
+            self._encode_rlite(vtx.get_id(), entry, hList, rdict)

```

**Comment by dongahn** [comment]
>Hi @dongahn, it wasn't clear to me if you wanted me to take a stab at fixing this issue, or were you going to work on both this and #893? Happy to help in any way.

Sorry I wasn't clear. I was going to look at both of the issues.

**Comment by grondo** [comment]
Great, thanks!
